### PR TITLE
Bump Dart SDK to `1.22.0-dev.10.1`.

### LINKED ISSUE #7736: Roll Dart SDK to 1.22.0-dev.10 (so we can format code with covariant)
Flutter is on Dart 1.22.0-dev.9.1 which has dart_style 0.2.14 which doesn't support the new `covariant`.

I'm told that if we roll to 1.22.0.dev.10 then we get dart_style 0.2.15 which does support `covariant`.

### PR DESCRIPTION
**Pull request by pq**
Fixes: #7736

Unblocks #7734 (migration to using `covariant` syntax).

@Hixie @abarth @tvolkert 

FYI @leafpetersen for strong mode suppressions

### CODE REVIEW

**Comment by tvolkert** [comment]
What's the story on the `strong_mode_not_instantiated_bound` suppressions?

File: packages/flutter_test/lib/src/controller.dart [hunk 1]
```diff
@@ -143,7 +143,7 @@ class WidgetController {
   ///
   /// * Use [firstState] if you expect to match several states but only want the first.
   /// * Use [stateList] if you expect to match several states and want all of them.
-  T state<T extends State>(Finder finder) {
+  T state<T extends State>(Finder finder) { // ignore: strong_mode_not_instantiated_bound
```
Comment (leafpetersen) [inline packages/flutter_test/lib/src/controller.dart]: There is a breaking change in the latest dev release to make strong mode be more conservative about guessing default type arguments when code leaves off a type argument.  Here, `State` is defined as

```dart
State<T extends StatefulWidget>
```

Previously, strong mode would just guess that you meant `State<StatefulWidget>` when you wrote `State`.  Writing this explicitly here will fix the error.

This error is primarily aimed at eliminating cases where we simply can't guess a reasonable answer, e.g.

```
class Foo<T extends Foo<T>> {}

Foo x; // What does Foo mean here?
```

There is some ongoing discussion as to whether to allow cases like `State` here where there is an answer we can guess.

@Hixie, @sethladd any thoughts about this?

**Comment by pq** [comment]
> What's the story on the strong_mode_not_instantiated_bound suppressions?

My understanding is that this warning is still a work in progress.  In any event @bwilkerson and I looked at a proper fix and I got shy since it would likely involve some invasive changes.

Maybe @leafpetersen can guide us to the light?

File: packages/flutter_tools/lib/src/ios/devices.dart [hunk 2]
```diff
@@ -278,7 +278,7 @@ class IOSDevice extends Device {
 
       Future<int> launch = runCommandAndStreamOutput(launchCommand, trace: true);
 
-      List<Uri> uris = await launch.then((int result) async {
+      List<Uri> uris = await launch.then<List<Uri>>((int result) async {
```
Comment (leafpetersen) [inline packages/flutter_tools/lib/src/ios/devices.dart]: Is this an inference regression, or an unrelated fix?

Comment (pq) [inline packages/flutter_tools/lib/src/ios/devices.dart]: Ah.  Good catch.  I should have flagged this.  Regression.  `flutter analyze` reported:

```
[error] Missing type arguments for generic method 'then<S>'. (packages/flutter_tools/lib/src/commands/daemon.dart, line 467, col 57)
[error] Missing type arguments for generic method 'then<S>'. (packages/flutter_tools/lib/src/ios/devices.dart, line 281, col 37)
```

File: packages/flutter_test/lib/src/controller.dart (hunk 1 above)
Comment (Hixie) [inline packages/flutter_test/lib/src/controller.dart]: State is marked `@optionalTypeArgs`. I would hope it would default to the broadest type possible (dynamic, or, in the case of a `<T extends Q>`, Q).

**Comment by Hixie** [comment]
Four new `// ignore`s will cost us $8012 (as measured by our technical debt benchmark).

File: packages/flutter_tools/lib/src/ios/devices.dart (hunk 2 above)
Comment (leafpetersen) [inline packages/flutter_tools/lib/src/ios/devices.dart]: Is there any chance that this is a holdover from dev.10.0?  We had a regression related to async/await, but I can't reproduce this error in either dev.10.1 or bleeding edge.

I'm not testing the actual code, so it's possible there's an issue I'm not seeing, but I have a repro that should be capturing everything relevant and I don't see any issues.

File: packages/flutter_test/lib/src/controller.dart (hunk 1 above)
Comment (leafpetersen) [inline packages/flutter_test/lib/src/controller.dart]: Ok, that's good to know.  In general that's what we do (that is, fill in with the bound), but we're being a bit more conservative with types used in bounds.

Would you expect to be able to use my example `Foo` above with no type arguments?  And if so what would you expect it to mean?

File: packages/flutter_tools/lib/src/ios/devices.dart (hunk 2 above)
Comment (pq) [inline packages/flutter_tools/lib/src/ios/devices.dart]: It was definitely failing `flutter analyze` w/ `dev.10.1` , though interestingly I was not seeing the errors in IntelliJ.

Maybe @danrubel can comment on that disconnect.  I haven't had time to follow-up on it...

File: packages/flutter_test/lib/src/controller.dart (hunk 1 above)
Comment (Hixie) [inline packages/flutter_test/lib/src/controller.dart]: I can't wrap my head around your example. I think I'd be ok with the compiler falling over when faced with that type. :-)

**Comment by pq** [comment]
Thanks for all the guidance @leafpetersen !

@Hixie : do we have a next-step?  It'd be great to land this so you all can start playing with covariants (right @sethladd 😉   ?)  Are there any tracking bugs I can add to the ignores to provide more context for future cleanup?  (Alternatively, I could link back to this PR I suppose...)

**Comment by sethladd** [comment]
Testing `covariant` is not a P0 for us, but when the new SDK lands for Flutter, we'll be happy to try it out.

**Comment by leafpetersen** [comment]
Filed https://github.com/dart-lang/sdk/issues/28580 to track the relaxation on the restriction on type bounds.

**Comment by goderbauer** [comment]
The latest engine version requires Dart SDK 1.22.0-dev.10.1. So, this is currently blocking the engine roll (https://github.com/flutter/flutter/pull/7777), which includes fixes for Flutter on Windows.

What is the path forward on bumping the Dart SDK?

**Comment by Hixie** [comment]
Closing this in favour of #7791.
